File: src/go/internal/storage/activity_start_gps_backfill.go
package storage

import (
	"context"
	"errors"
	"fmt"
	"time"

	"gorm.io/gorm"
)

// ActivityStartGPSBackfillOptions controls the one-off activity start cache
// backfill. Limit is a total activity cap for validation runs; zero is unlimited.
type ActivityStartGPSBackfillOptions struct {
	UserIDs   []string
	Commit    bool
	Limit     int
	BatchSize int
	Delay     time.Duration
}

// Validate rejects unsafe backfill settings before any database query runs.
func (o ActivityStartGPSBackfillOptions) Validate() error {
	if len(o.UserIDs) == 0 {
		return errors.New("activity start GPS backfill requires at least one user")
	}
	for _, userID := range o.UserIDs {
		if _, err := canonicalUserID(userID); err != nil {
			return err
		}
	}
	if o.BatchSize < 1 || o.BatchSize > 500 {
		return errors.New("activity start GPS backfill batch size must be from 1 to 500")
	}
	if o.Delay < 0 {
		return errors.New("activity start GPS backfill delay must be non-negative")
	}
	if o.Limit < 0 {
		return errors.New("activity start GPS backfill limit must be non-negative")
	}
	return nil
}

// ActivityStartGPSBackfillReport contains counts only; it deliberately omits
// user IDs, activity IDs, coordinates, and database configuration.
type ActivityStartGPSBackfillReport struct {
	Mode              string `json:"mode"`
	Users             int    `json:"users"`
	Total             int64  `json:"total"`
	AlreadyCached     int64  `json:"already_cached"`
	Missing           int64  `json:"missing"`
	Scanned           int    `json:"scanned"`
	Fillable          int    `json:"fillable"`
	Updated           int    `json:"updated"`
	Verified          int    `json:"verified"`
	NoValidGPS        int    `json:"no_valid_gps"`
	SkippedConcurrent int    `json:"skipped_concurrent"`
	Failed            int64  `json:"failed"`
}

type activityStartGPSCounts struct {
	Total   int64
	Cached  int64
	Missing int64
	Partial int64
}

type activityStartGPSPoint struct {
	GPSLat float64
	GPSLon float64
}

// BackfillActivityStartGPS fills activities.start_gps_* from the first complete,
// valid timeseries coordinate. Activities are keyset-paged; each selected row
// performs one lookup constrained by the (user_id, label_id) index.
func (s *Store) BackfillActivityStartGPS(ctx context.Context, options ActivityStartGPSBackfillOptions) (ActivityStartGPSBackfillReport, error) {
	if err := options.Validate(); err != nil {
		return ActivityStartGPSBackfillReport{}, err
	}
	report := ActivityStartGPSBackfillReport{Mode: "dry-run"}
	if options.Commit {
		report.Mode = "commit"
	}

	for _, userID := range options.UserIDs {
		uid, _ := canonicalUserID(userID) // Validate already checked every ID.
		report.Users++
		counts, err := s.countActivityStartGPS(ctx, uid)
		if err != nil {
			return report, fmt.Errorf("storage: count activity start GPS rows: %w", err)
		}
		report.Total += counts.Total
		report.AlreadyCached += counts.Cached
		report.Missing += counts.Missing
		report.Failed += counts.Partial

		cursor := ""
		for options.Limit == 0 || report.Scanned < options.Limit {
			pageSize := options.BatchSize
			if options.Limit > 0 && pageSize > options.Limit-report.Scanned {
				pageSize = options.Limit - report.Scanned
			}
			labels, err := s.missingActivityStartGPSLabels(ctx, uid, cursor, pageSize)
			if err != nil {
				return report, fmt.Errorf("storage: page activity start GPS rows: %w", err)
			}
			if len(labels) == 0 {
				break
			}
			for _, labelID := range labels {
				cursor = labelID
				report.Scanned++
				if err := s.backfillOneActivityStartGPS(ctx, uid, labelID, options.Commit, &report); err != nil {
					report.Failed++
				}
				if options.Delay > 0 {
					select {
					case <-ctx.Done():
						return report, ctx.Err()
					case <-time.After(options.Delay):
					}
				}
			}
		}
	}
	return report, nil
}

func (s *Store) countActivityStartGPS(ctx context.Context, userID string) (activityStartGPSCounts, error) {
	var counts activityStartGPSCounts
	err := s.db.WithContext(ctx).Model(&Activity{}).
		Select(`COUNT(*) AS total,
COALESCE(SUM(start_gps_lat IS NOT NULL AND start_gps_lon IS NOT NULL), 0) AS cached,
COALESCE(SUM(start_gps_lat IS NULL AND start_gps_lon IS NULL), 0) AS missing,
COALESCE(SUM((start_gps_lat IS NULL) <> (start_gps_lon IS NULL)), 0) AS partial`).
		Where("user_id = ?", userID).Scan(&counts).Error
	return counts, err
}

func (s *Store) missingActivityStartGPSLabels(ctx context.Context, userID, cursor string, limit int) ([]string, error) {
	var labels []string
	err := s.db.WithContext(ctx).Model(&Activity{}).Select("label_id").
		Where("user_id = ? AND label_id > ?", userID, cursor).
		Where("start_gps_lat IS NULL AND start_gps_lon IS NULL").
		Order("label_id").Limit(limit).Find(&labels).Error
	return labels, err
}

func (s *Store) backfillOneActivityStartGPS(ctx context.Context, userID, labelID string, commit bool, report *ActivityStartGPSBackfillReport) error {
	var point activityStartGPSPoint
	err := s.db.WithContext(ctx).Table("timeseries FORCE INDEX (idx_timeseries_user_label)").
		Select("gps_lat, gps_lon").
		Where("user_id = ? AND label_id = ?", userID, labelID).
		Where("gps_lat IS NOT NULL AND gps_lon IS NOT NULL").
		Where("gps_lat BETWEEN -90 AND 90 AND gps_lon BETWEEN -180 AND 180").
		Where("NOT (gps_lat = 0 AND gps_lon = 0)").
		Order("id").Take(&point).Error
	if errors.Is(err, gorm.ErrRecordNotFound) {
		report.NoValidGPS++
		return nil
	}
	if err != nil {
		return err
	}
	report.Fillable++
	if !commit {
		return nil
	}

	result := s.db.WithContext(ctx).Model(&Activity{}).
		Where("user_id = ? AND label_id = ?", userID, labelID).
		Where("start_gps_lat IS NULL AND start_gps_lon IS NULL").
		Updates(map[string]any{"start_gps_lat": point.GPSLat, "start_gps_lon": point.GPSLon})
	if result.Error != nil {
		return result.Error
	}
	if result.RowsAffected == 0 {
		report.SkippedConcurrent++
		return nil
	}
	report.Updated++

	var stored Activity
	if err := s.db.WithContext(ctx).Select("start_gps_lat, start_gps_lon").
		Where("user_id = ? AND label_id = ?", userID, labelID).Take(&stored).Error; err != nil {
		return err
	}
	if stored.StartGPSLat == nil || stored.StartGPSLon == nil ||
		*stored.StartGPSLat != point.GPSLat || *stored.StartGPSLon != point.GPSLon {
		return errors.New("activity start GPS readback mismatch")
	}
	report.Verified++
	return nil
}

File: src/migration/src/activity-start-gps-backfill.js
#!/usr/bin/env node
// Backfill activities.start_gps_* from each activity's first valid timeseries
// GPS pair. The CLI is dry-run by default and processes one indexed activity
// lookup at a time to avoid recreating the all-history GROUP BY load spike.

import { readFileSync } from "node:fs";
import { dirname, join } from "node:path";
import { fileURLToPath } from "node:url";
import { parseArgs } from "node:util";

import { connect, parseMysqlConfig } from "./mysql.js";
import { selectUserIds } from "./profiles.js";
import { users as REAL_USERS } from "./users.js";

const PROJECT_DIR = join(dirname(fileURLToPath(import.meta.url)), "..");

const ACTIVITY_COUNTS_SQL = `
SELECT
  COUNT(*) AS total,
  COALESCE(SUM(start_gps_lat IS NOT NULL AND start_gps_lon IS NOT NULL), 0) AS cached,
  COALESCE(SUM(start_gps_lat IS NULL AND start_gps_lon IS NULL), 0) AS missing,
  COALESCE(SUM((start_gps_lat IS NULL) <> (start_gps_lon IS NULL)), 0) AS partial
FROM activities
WHERE user_id = ?`;

function missingActivitiesSQL(limit) {
  return `
SELECT label_id
FROM activities
WHERE user_id = ? AND label_id > ?
  AND start_gps_lat IS NULL AND start_gps_lon IS NULL
ORDER BY label_id
LIMIT ${limit}`;
}

const FIRST_VALID_GPS_SQL = `
SELECT gps_lat, gps_lon
FROM timeseries FORCE INDEX (idx_timeseries_user_label)
WHERE user_id = ? AND label_id = ?
  AND gps_lat IS NOT NULL AND gps_lon IS NOT NULL
  AND gps_lat BETWEEN -90 AND 90 AND gps_lon BETWEEN -180 AND 180
  AND NOT (gps_lat = 0 AND gps_lon = 0)
ORDER BY id
LIMIT 1`;

const UPDATE_ACTIVITY_START_SQL = `
UPDATE activities
SET start_gps_lat = ?, start_gps_lon = ?
WHERE user_id = ? AND label_id = ?
  AND start_gps_lat IS NULL AND start_gps_lon IS NULL`;

const VERIFY_ACTIVITY_START_SQL = `
SELECT start_gps_lat, start_gps_lon
FROM activities
WHERE user_id = ? AND label_id = ?`;

function number(row, key) {
  return Number(row?.[key] ?? 0);
}

function emptyReport(commit) {
  return {
    mode: commit ? "commit" : "dry-run",
    users: 0,
    total: 0,
    already_cached: 0,
    missing: 0,
    scanned: 0,
    fillable: 0,
    updated: 0,
    verified: 0,
    no_valid_gps: 0,
    skipped_concurrent: 0,
    failed: 0,
  };
}

export async function runActivityStartGPSBackfill({
  connection,
  userIds,
  commit = false,
  batchSize = 25,
  delayMs = 25,
  maxActivities = Infinity,
}) {
  if (!connection) throw new Error("MySQL connection is required");
  if (!Number.isInteger(batchSize) || batchSize < 1 || batchSize > 500) {
    throw new Error("batchSize must be an integer from 1 to 500");
  }
  if (!Number.isFinite(delayMs) || delayMs < 0) {
    throw new Error("delayMs must be non-negative");
  }
  if (maxActivities !== Infinity && (!Number.isInteger(maxActivities) || maxActivities < 1)) {
    throw new Error("maxActivities must be a positive integer");
  }
  const report = emptyReport(commit);
  for (const userId of userIds) {
    report.users++;
    const [countRows] = await connection.execute(ACTIVITY_COUNTS_SQL, [userId]);
    const counts = countRows[0] || {};
    report.total += number(counts, "total");
    report.already_cached += number(counts, "cached");
    report.missing += number(counts, "missing");
    report.failed += number(counts, "partial");

    let cursor = "";
    while (report.scanned < maxActivities) {
      const limit = Math.min(batchSize, maxActivities - report.scanned);
      const [activities] = await connection.execute(missingActivitiesSQL(limit), [
        userId, cursor,
      ]);
      if (activities.length === 0) break;
      for (const activity of activities) {
        cursor = activity.label_id;
        report.scanned++;
        try {
          const [points] = await connection.execute(FIRST_VALID_GPS_SQL, [
            userId, activity.label_id,
          ]);
          if (points.length === 0) {
            report.no_valid_gps++;
          } else {
            report.fillable++;
            if (commit) {
              const point = points[0];
              const [result] = await connection.execute(UPDATE_ACTIVITY_START_SQL, [
                point.gps_lat, point.gps_lon, userId, activity.label_id,
              ]);
              if (result.affectedRows === 0) {
                report.skipped_concurrent++;
              } else {
                report.updated++;
                const [rows] = await connection.execute(VERIFY_ACTIVITY_START_SQL, [
                  userId, activity.label_id,
                ]);
                const stored = rows[0];
                if (
                  !stored || Number(stored.start_gps_lat) !== Number(point.gps_lat) ||
                  Number(stored.start_gps_lon) !== Number(point.gps_lon)
                ) {
                  throw new Error("activity start GPS readback mismatch");
                }
                report.verified++;
              }
            }
          }
        } catch {
          report.failed++;
        }
        if (delayMs > 0) await new Promise((resolve) => setTimeout(resolve, delayMs));
      }
    }
  }
  return report;
}

function loadDotEnv(dir, env) {
  for (const file of [".env", ".env.local"]) {
    let text;
    try {
      text = readFileSync(join(dir, file), "utf8");
    } catch {
      continue;
    }
    for (const raw of text.split(/\r?\n/)) {
      const line = raw.trim();
      if (!line || line.startsWith("#")) continue;
      const eq = line.indexOf("=");
      if (eq < 0) continue;
      const key = line.slice(0, eq).trim();
      let value = line.slice(eq + 1).trim();
      if ((value.startsWith('"') && value.endsWith('"')) ||
          (value.startsWith("'") && value.endsWith("'"))) {
        value = value.slice(1, -1);
      }
      if (!(key in env)) env[key] = value;
    }
  }
}

export function parseActivityStartGPSCli(argv) {
  const { values } = parseArgs({
    args: argv,
    options: {
      commit: { type: "boolean", default: false },
      user: { type: "string", multiple: true, default: [] },
      limit: { type: "string" },
      "batch-size": { type: "string", default: "25" },
      "delay-ms": { type: "string", default: "25" },
      help: { type: "boolean", default: false },
    },
    allowPositionals: false,
  });
  const maxActivities = values.limit == null ? Infinity : Number(values.limit);
  const batchSize = Number(values["batch-size"]);
  const delayMs = Number(values["delay-ms"]);
  if (maxActivities !== Infinity && (!Number.isInteger(maxActivities) || maxActivities < 1)) {
    throw new Error("--limit must be a positive integer");
  }
  if (!Number.isInteger(batchSize) || batchSize < 1 || batchSize > 500) {
    throw new Error("--batch-size must be an integer from 1 to 500");
  }
  if (!Number.isInteger(delayMs) || delayMs < 0) {
    throw new Error("--delay-ms must be a non-negative integer");
  }
  return {
    commit: values.commit,
    requestedUsers: values.user.flatMap((value) => value.split(","))
      .map((value) => value.trim().toLowerCase()).filter(Boolean),
    maxActivities, batchSize, delayMs, help: values.help,
  };
}

function usage() {
  process.stdout.write("backfill-activity-start-gps - timeseries to activities.start_gps_*\n\n" +
    "Usage: npm run migrate:activity-start-gps -- [options]\n\n" +
    "  --commit           Write and verify cached starts. Default is dry-run.\n" +
    "  --user <uuid>      Real-user UUID from src/users.js; repeatable/comma-separated.\n" +
    "  --limit <n>        Scan at most n missing activities across selected users.\n" +
    "  --batch-size <n>   Keyset page size, 1-500 (default 25).\n" +
    "  --delay-ms <n>     Delay after every activity lookup (default 25ms).\n" +
    "  --help             Show this help.\n\n" +
    "MySQL uses STRIDE_WORKER_MYSQL_DSN or the discrete MYSQL_* variables.\n" +
    "Only real users declared in src/users.js are eligible. Output contains counts only.\n");
}

export async function main(argv = process.argv.slice(2), env = process.env) {
  loadDotEnv(PROJECT_DIR, env);
  const options = parseActivityStartGPSCli(argv);
  if (options.help) {
    usage();
    return 0;
  }
  const { ids, rejected } = selectUserIds(REAL_USERS, options.requestedUsers);
  if (rejected.length > 0) {
    throw new Error("--user is not in src/users.js real-user allowlist: " + rejected.join(","));
  }
  if (ids.length === 0) throw new Error("no real users selected");
  const connection = await connect(parseMysqlConfig(env));
  try {
    const report = await runActivityStartGPSBackfill({
      connection, userIds: ids, commit: options.commit,
      batchSize: options.batchSize, delayMs: options.delayMs,
      maxActivities: options.maxActivities,
    });
    process.stdout.write(JSON.stringify(report) + "\n");
    return report.failed > 0 ? 1 : 0;
  } finally {
    await connection.end();
  }
}

if (process.argv[1] === fileURLToPath(import.meta.url)) {
  main().then((code) => { process.exitCode = code; }).catch((error) => {
    process.stderr.write("fatal: " + (error?.message || "migration failed") + "\n");
    process.exitCode = 2;
  });
}
